TypeCobol/Compiler/CupParser/NodeBuilder/ProgramClassBuilder.cs

```diff
@@ -230,12 +231,49 @@ public virtual void StartCobolProgram(ProgramIdentification programIdentificatio
             Dispatcher.StartCobolProgram(programIdentification, libraryCopy);
         }
 
-        public virtual void EndCobolProgram(TypeCobol.Compiler.CodeElements.ProgramEnd end)
+        public virtual void EndCobolProgram(ProgramEnd end)
         {
-            AttachEndIfExists(end);
-            Exit();
-            programsStack.Pop();
-            Dispatcher.EndCobolProgram(end);
+            if (end != null)
+            {
+                if (end.ProgramName?.Name != null)
+                {
+                    var matchingProgram = FindMatchingProgram();
+                    if (matchingProgram != null)
+                    {
+                        // Pop all programs to have CurrentProgram as the program whose name match
+                        while (matchingProgram != CurrentProgram)
+                        {
+                            // Skip closing all nested programs
+                            Exit();
+                            programsStack.Pop();
+                            Dispatcher.EndCobolProgram(null);
+                        }
+                    }
+                }
+            }
+            else
+            {
+                // When end is null, it is has been automatically generated
+                if (CurrentProgram?.Children.LastOrDefault() is End endNode && endNode.CodeElement.Type == CodeElementType.ProgramEnd)
+                {
+                    // It can be useless because it was generated when we already closed the CurrentProgram
+                    return;
+                }
+            }
+
+            if (CurrentProgram != null)
```

Comment:
It isn't a `Debug.Assert` because we could have closed all programs manually but the parser still wants to call us. It would result in a bug.
For example in this case, when meeting `END PROGRAM Stacked0.`, we close Nested0 manually, and then Stacked0 with the encountered end.
But the parser also calls the method one more time because it assumed `END PROGRAM Stacked0.` closed Nested0 and it's next call was a try to correct the program by closing Stacked0.
```cobol
       IDENTIFICATION DIVISION.
       PROGRAM-ID. Stacked0.
      
       IDENTIFICATION DIVISION.
       PROGRAM-ID. Nested0.
      
       END PROGRAM Stacked0.
```